src/extension/nodes/CompoundCloudPlane.cpp

```diff
@@ -0,0 +1,479 @@
+    ClassDB::bind_method(D_METHOD("convert_to_world", "cloud_x", "cloud_y"), &CompoundCloudPlane::convert_to_world);
+
+    ClassDB::bind_method(D_METHOD("get_compounds"), &CompoundCloudPlane::get_compounds);
+    ClassDB::bind_method(D_METHOD("set_compounds", "compounds"), &CompoundCloudPlane::set_compounds);
+    ClassDB::bind_method(D_METHOD("get_resolution"), &CompoundCloudPlane::get_resolution);
+    ClassDB::bind_method(D_METHOD("set_resolution", "resolution"), &CompoundCloudPlane::set_resolution);
+    ClassDB::bind_method(D_METHOD("get_size"), &CompoundCloudPlane::get_size);
+    ClassDB::bind_method(D_METHOD("set_size", "size"), &CompoundCloudPlane::set_size);
+    ClassDB::bind_method(D_METHOD("get_compounds_array"), &CompoundCloudPlane::get_compounds_array);
+    ClassDB::bind_method(D_METHOD("get_compounds_count"), &CompoundCloudPlane::get_compounds_count);
+
+    ADD_PROPERTY(PropertyInfo(Variant::PACKED_INT32_ARRAY, "compounds"), "set_compounds", "get_compounds");
+    ADD_PROPERTY(PropertyInfo(Variant::INT, "resolution"), "set_resolution", "get_resolution");
+    ADD_PROPERTY(PropertyInfo(Variant::INT, "size"), "set_size", "get_size");
+}
+
+CompoundCloudPlane::CompoundCloudPlane() {
+    instance_id = next_instance_id++;
+    instances[instance_id] = this;
+    resolution = 0;
+    size = 0;
+    brightness = 1.0f;
+    position = godot::Vector2(0, 0);
+    texture_needs_update = false;
+    minimum_visible_density = 0.01f;
+
+    for (int i = 0; i < CLOUDS_IN_ONE; ++i) {
+        decay_rates[i] = 0.99f - (i * 0.01f);
+        compound_colors[i] = godot::Color(1, 1, 1, 1);  
+    }
+}
+
+CompoundCloudPlane::~CompoundCloudPlane() {
+    instances.erase(instance_id);
+}
+
+void CompoundCloudPlane::_ready() {
+    create_mesh();
+    create_shader_material();
+    create_noise_texture();
+    if (size > 0) {
+        update_cloud_texture();
+    }
+    update_shader_parameters();
+
+    cached_world_position = get_global_transform().origin;
+    set_process(false);  // Start with processing disabled
+}
+
+void CompoundCloudPlane::_process(double delta) {
+    diffuse(static_cast<float>(delta));
+    advect(static_cast<float>(delta));
+    
+    if (texture_needs_update) {
+        update_cloud_texture();
+        texture_needs_update = false;
+    }
+    
+    update_shader_parameters();
+
+    if (!has_compounds()) {
+        set_process(false);
+    }
+}
+
+int64_t CompoundCloudPlane::get_native_instance() {
+    return instance_id;
+}
+
+void CompoundCloudPlane::init(int render_priority, int cloud1, int cloud2, int cloud3, int cloud4) {
+    compounds.resize(4);
+    compounds.set(0, cloud1);
+    compounds.set(1, cloud2);
+    compounds.set(2, cloud3);
+    compounds.set(3, cloud4);
+
+    set_process_priority(render_priority);
+
+    resolution = 84;
+    size = 104;
+
+    if (size <= 0) {
+        godot::UtilityFunctions::printerr("Invalid size for CompoundCloudPlane");

```

Comment:
The helper error macro should be used instead as that will automatically include the method name and line number the error happens so that's much more helpful way to print errors.